### ISSUE TITLE
Remove partial copy-on-write mechanism of ParameterMap in ExperimentInfo

### ISSUE THREAD

**Issue by SimonHeybrock** [mantidproject/mantid#19001]
`ExperimentInfo::instrumentParameters()` has a copy-on-write mechanism: It copies the `ParameterMap` if the reference to it is not unique.

There are several reasons for non-unique references:

1. Client code has obtained a `shared_ptr<Instrument>` or a `shared_ptr<IDetector>` from the workspaces. `ExperimentInfo::instrumentParameters()` will then trigger a copy. This implies the following behavior:

  ```cpp
  const auto det = ws.getInstrument()->getDetector(42);
  MoveInstrumentComponent(ws, 42, oldPos + offset); // not actual code
  ws.getInstrument()->getDetector(42)->getPos(); // oldPos + offset
  det->getPos(); // oldPos
  ```

  This is probably unexpected, and there is little to no use of this functionality in Mantid.

2. Several workspaces sharing the same `ParameterMap`. This was possible until recently, except for `MatrixWorkspace`, which explicitly triggered a copy in its constructor. After a [recent change](https://github.com/mantidproject/mantid/pull/18831) a copy is always made, so this is not possible anymore.

3. `DetectorInfo` holds a reference to the `ParameterMap` (directly or indirectly). Conceptually this is similar to item (1.), so this sharing does not require a copy.

In conclusion, since we have https://github.com/mantidproject/mantid/pull/18831, it should be possible to remove the COW mechanism, if we fix potential code that relies on the functionality described in (1.).

There are two important benefits:

- Calling `ExperimentInfo::instrumentParameters()` will no longer invalidate references to `SpectrumInfo` or `DetectorInfo`. Several developers working on the rollout during the maintenance period suffered from issues due to this unfortunate behavior.
- `DetectorInfo` will be flushed much less frequently, and this the moderately costly rebuilds can be avoided.
